calyx-py/test/arb.py

```diff
@@ -0,0 +1,136 @@
+# pylint: disable=import-error
+from calyx.py_ast import Stdlib, CompPort, CompVar, ParComp, Enable, If
+import calyx.builder as cb
+
+# AM, quality of life:
+# It occurs to me that a nice goal for the builder library folks
+# could be to introduce enough functionality that we don't need to
+# import anything from calyx.py_ast.
+
+
+def add_eq(comp, port_name, const, cellname, groupname):
+    """Adds wiring to check `port_name == const`,
+    where `port_name` is a port and `const` is an integer constant."""
+    eq_cell = comp.cell(cellname, Stdlib.op("eq", 32, signed=False))
+    with comp.comb_group(groupname):
+        eq_cell.left = CompPort(CompVar(port_name), "out")
+        # AM, point of failure:
+        # This is wrong. It renders "{port_name}.out" but I want "{port_name}".
+        # The same issue occurs repeatedly, so I'm just flagging this one.
+        # Whenever the generated futil has {i/j}.out, I actually just want {i/j}.
+        eq_cell.right = cb.const(32, const)
+
+
+def add_emit_from_mem(comp, mem, ans, suffix):
+    """Adds wiring that puts mem{suffix}[j] into ans."""
+    with comp.group(f"emit_from_mem{suffix}") as emit_from_mem:
+        mem.addr0 = CompPort(CompVar("j"), "out")  # AM: want j, not j.out
+        ans.write_en = 1
+        ans.write_data = mem.read_data
+        emit_from_mem.done = ans.done
+
+
+def add_wrap(prog):
+    """Inserts the component `wrap` into the program.
+
+    It has:
+    - two inputs, `i` and `j`
+    - two ref memories, `mem1` and `mem2`
+    - one output, `out`
+
+    For now, assume 0 <= i < 2 and 0 <= j < 4.
+    if i == 0, then out = mem1[j]
+    if i == 1, then out = mem2[j]
+    """
+
+    wrap = prog.component("wrap")
+    wrap.input("i", 32)
+    wrap.input("j", 32)
+
+    mem1 = wrap.mem_d1("mem1", 32, 4, 32, is_ref=True)
+    mem2 = wrap.mem_d1("mem2", 32, 4, 32, is_ref=True)
+    ans = wrap.mem_d1("ans", 32, 1, 32, is_ref=True)
+
+    add_eq(wrap, "i", 0, "eq0", "i_eq_0")
+    add_eq(wrap, "i", 1, "eq1", "i_eq_1")
+    add_emit_from_mem(wrap, mem1, ans, "1")
+    add_emit_from_mem(wrap, mem2, ans, "2")
+
+    # AM, quality of life:
+    # I'd like to generate these `if` statements with the builder.
```

Comment:
Useful to point this out to @EclecticGriffin as a potential `builder` improvement

Reply:
FWIW, it should be possible to do this with `cb.if_` when using the values returned from building the stuff. That is, this should hypothetically work:

```py
cb.if_(eq0.out, i_eq_0, emit_from_mem1)
```

But that will require making, for example, the `add_eq` function return a value (`eq_cell`) and putting that result into the variable `eq0`. But that should simplify things a bit vs. using the string names everywhere.